Added --exclude feature

xIceFox opened the pull request:
I was missing an option to exclude projects.
This is especially useful if you want to build only the affected projects, but no projects which are including tests.

To accomplish this, I added the command line option --exclude, where you can specify a regulary expression, which filters out all project names that are matching this expression.

xIceFox commented:
@leonardochaia Tell me what you think 😄 
I tried to match the projects coding style etc.

leonardochaia commented:
Hi @xIceFox , thanks for the PR. I haven't looked at the code yet but wanted to ask if you've seen [DotnetAffected.Tasks](https://github.com/leonardochaia/dotnet-affected#sdk-based-installation) yet?

It is for advanced use-cases, I am personally not using it yet but hoping to do so soon. [This are the docs](https://github.com/leonardochaia/dotnet-affected/blob/main/src/DotnetAffected.Tasks/README.md)
One of the features, is [Context Filtering](https://github.com/leonardochaia/dotnet-affected/blob/main/src/DotnetAffected.Tasks/README.md#context-filtering) which lets you filter based on MSBuild properties.

Since Dotnet.Affected.Tasks is an MSBuild SDK, it can be composed with other SDKs like traversal and you can simply use `<ProjectReference Remove="..."` to exclude projects, [docs](https://github.com/leonardochaia/dotnet-affected/blob/main/src/DotnetAffected.Tasks/README.md#extensibility)

So, I think the CLI needs to have a way to filter projects but I haven't thought about it yet, but I guess it should be as similar as possible as Dotnet.Affected.Tasks.

I'll take a look later today :+1:

leonardochaia commented:
Yup, I agree the CLI needs to have the same features, I just wanna make sure that if we have same features on CLI/Tasks they should both behave the same and using the same names/wording, etc.

I'll take a look later today :+1:

xIceFox commented:
@leonardochaia Cool, I will give it a try. Maybe it will be a few days, because in need to read me into the core code. With .NET Regex you mean the dependency I used in this PR?

leonardochaia commented:
Hi @xIceFox , sure thing take you time  :+1:

Yeah I meant `System.Regex`  class, perhaps AffectedOptions can receive the regex as string in constructor and instantiate the System.Regex there.

I also need to take a look at the code to see how to fit the exclude but I do think it should go into Core so that it can be reused by higher order libraries like the CLI and Tasks.

xIceFox commented:
Hi @leonardochaia.
I implemented it.
Don't be suprised by the amout of files changed, i changed the namespace of the Core Extensions to have a clear seperation between the dependencies.
I think the my solution can be improved, so if you find something that bothers you just write me.

xIceFox commented:
My changes broke something in DotnetAffected.Tasks, I need to find out why.

xIceFox commented:
I think i found the problem, it was a missing parameter in AffectedOptions, but the tests are still failing on my computer, because everywhere in the tests there are missing quotation marks in paths, therefore paths with spaces create errors. I'm trying to fix that now.

xIceFox commented:
Tests work now.

leonardochaia commented:
Hi @xIceFox , thanks for the changes, gimme a few days to review. I'll most likely cherry pick what I need. Unfortunately, changing namespaces is a breaking change on the public API, and to be frank namespaces were chosen deliberately. I do want the exclusion feature so let me take a look and see what we can do :+1: 

For the future, when possible please create PRs with separate features instead of a single PR with multiple features/refactors etc. :+1:  
That simplifies reviewing, but also merging, since I can merge faster what is ready and if something needs changes it does not hold on on the other unrelated changes.

Personally, I use git machete to manage multiple dependent PR branches.

Thanks again for your contribution @xIceFox :1st_place_medal:

xIceFox commented:
Hi @leonardochaia,
That is totally fine for me.
Have a great one ☺️
